docs: Add component design guidelines (#14265)
backstage/backstage: docs/dls/component-design-guidelines.md

---
id: component-design-guidelines
title: Component Design Guidelines
description: Documentation on Design
---

Be it a new component contribution, or plugin specific components, you'll want

> to follow these guidelines, we'll cover the 3 main subjects that define the

Reviewer: ```suggestion
to follow these guidelines. We'll cover the three main subjects that define the
```

> general look and feel of your components, all of which build on top of the MUI

Reviewer: ```suggestion
general look and feel of your components, all of which build on top of the Material-UI
```

theme features:

- Layout
- Color palette
- Typography

## 🏗️ Layout

> Layout refers to how you organize or stack content, whenever possible we want

Reviewer: ```suggestion
Layout refers to how you organize or stack content. Whenever possible, we want
```

to use Backstage's components (check the [Storybook][1] for a list and demo)

> or MUI components (check the [MUI docs][2]).

Reviewer: ```suggestion
first, and otherwise fall back to Material-UI components (check the [MUI docs][2]).
```

> If none of these fit your layout needs, then you can build one, however using

Reviewer: ```suggestion
If none of these fit your layout needs, then you can build your own components. However, using
```

> HTML+CSS directly is not recommended, it's better to use MUI layout components

Reviewer: ```suggestion
HTML+CSS directly is not recommended: it's better to use MUI layout components
```

Author: Not sure if this one makes sense, I'm not a native english speaker but is putting a colon `:` in there syntactically correct?

Author: Changed, let me know if `:` is better than `;`

to make your layout theme aware, meaning if someone changes the theme, your
layout would react to those changes without requiring updates to your code.

Specifically you want to look at these components that make use of the
`theme.spacing()` function for margins, paddings and positions, as well as
color palette and typography:

- [Container][3] mostly at page level
- [Box][4] like a div that can be customized a lot

> - [Grid][5] & [Grid V2][6] (preferable V2) for flexible grid layouts

Reviewer: I think you'll want to look through these links and all of the other ones below too. We are on MUI4, so all of the links should be to the version 4 docs and at least Grid2 isn't there at all.

Author: I didn't want to put the V4 explicitly in case we update Backstage to latest and we forget to update the docs here, in this form it would always point to latest MUI, should I change it regardless?

Author: Changed to V4

- [Stack][7] for vertical layouts (single column)
- [Paper][8] The base of a card, like it's background & padding on the borders
- [Card][9] Card with support for title, description, buttons, images...

> ## Color palette

Reviewer: Regarding this section, there's [already this](https://backstage.io/docs/getting-started/app-custom-theme) which explains how to change overall theming - which is the preferred practice to reach for first, before doing styling on individual things (which of course you sometimes want to do as well)

Author: Want me to link that and leave all this explanation as the last resort?

Author: Did changes on this section.

Any component that needs a color to put in the styles should be using the

> theme's color palette, most Backstage components and all MUI components should

Reviewer: ```suggestion
theme's color palette. Most Backstage components and all MUI components should
```

use the theme's color palette by default, so unless you need explicit control

> on the color of a component (say when the component was design to use the

Reviewer 2: ```suggestion
on the color of a component (say when the component was designed to use the
```

primary color but you want to use the secondary), then the easiest way to access
the color palette is with [`useTheme` hook](https://mui.com/material-ui/customization/theming/#accessing-the-theme-in-a-component).

It's not a very common use case to override a theme color in a MUI component

> but let's say you have a Paper component that highlights it's content with a

Reviewer: ```suggestion
but let's say you have a Paper component that highlights its content with a
```

different color for a side menu or something (usually you use the elevation,

> but maybe the designer wanted a colorful app), you can use the theme like this:

Reviewer: ```suggestion
but maybe the designer wanted a colorful app). You can use the theme like this:
```

```tsx
import { useTheme } from '@mui/material/styles';

export function Sidebar() {
  const theme = useTheme();
  return (
    <Paper

>       sx={{

Reviewer: mui4 Paper doesn't have an `sx` prop - and we usually go via makeStyles and classes to adapt things. E.g.: https://github.com/backstage/backstage/blob/7539b3674830af1ca98ea3d74ca4be6b290c4473/packages/create-app/templates/default-app/packages/app/src/components/search/SearchPage.tsx#L55

Author: Indeed, that is the case for MUI 4, I'll change this to use `makeStyles` and classes instead.

Author: Done.

        backgroundColor: theme.palette.primary.main,
      }}
    >
      Some children here
    </Paper>
  );
}
```

Here is a link to the [Default Palette values][10] you can use, the tokens will be the same, what changes are the colors associated with those depending on
your app theme color palette.

## Typography

Most of the time the components from MUI will use the `<Typography />` component
which will use the theme's typography properties like font family, size, weight

> and appropriate color from the palette for the context of that component, like

Reviewer: ```suggestion
and appropriate color from the palette for the context of that component. This applies for example to 
```
